### ISSUE TITLE
Take ownership process doesn't look complete 

### ISSUE THREAD

**Issue by alexeyzimarev** [AsynkronIT/protoactor-dotnet#530]
When the new cluster member joins, the existing nodes try to rebalance the load by transferring ownership to the new node.

However, the process is fire and forget, the ownership transfer has no confirmation.

Also, actors that have been transferred are only removed from the partition but stay active in the process registry and in the PidCache. When attempting to get an actor on the same node, it will be returned from the PidCache, pointing to the current node. I think it might lead to the same actor being split and get to a split-brain problem and inconsistent state.

I'd say we need to stop transferred actors and by doing so wipe them from the PidCache, forcing the next call to `GetActorAsync` to ask the partition for the new location of the actor.

**Comment by alexeyzimarev** [comment]
Another thing is that I might be missing something important here, but when the ownership transfer message gets to the new node, it just adds the PID to the partition. But it doesn't spawn the actor. When I try getting n actor by name and kind from the cluster on that node, the PID is returned by the partition but there's no actor.

**Comment by alexeyzimarev** [comment]
Ok, I was tracing the issue and discovered the following:

- we send transfer ownership messages to the other node
- the other node puts the actor in its partition
- the sending node also moves it to the remote node partition
- that node goes down permanently

As a result, zombie addresses are present on the remaining nodes in PidCache and in the partitions list for that non-existent member and all attempts to talk to those actors naturally fail.

Removing the whole ownership transfer process and replacing it with the actor stop, partition cleanup and pid cache cleanup on the node that transfers the ownership from, solves the issue. Actors disappear from the node and requests to the actor triggers the cluster requesting the remote actor from the correct node.

**Comment by alexeyzimarev** [comment]
Now I managed to remove zombie actors from partitions but some of them still hang in the pid cache on some nodes. Need to investigate more.

**Comment by alexeyzimarev** [comment]
The solution that works in my setup now includes:

- Do now use the `TakeOwnership` message at all, don't send it and don't handle it in the `PartitionActor`
- Change the `PartitionActor.TransferOwnership` to
  - stop the transferred actor instead of unwatching it
  - remove the transferred actor `PID` from the cluster `PidCache`

After applying these changes, I am able to run the cluster, increase and decrease the number of nodes with the cluster rebalancing itself. Senders that use `Cluster.GetActorAsync` are getting the new remote address of the actor after it has been removed from the current node.

Before those changes, I was monitoring the content of the partitions list and actors for each partition, and the cluster pid cache. I observed that actors that should've been taken by another node remained in the pid cache. Also, partitions were kept for dead nodes because there was no `MemberLeftEvent` for the node that is taken down at the same time as the new node starts (Kubernetes deployment replica set scenario).

**Comment by alexeyzimarev** [comment]
I guess the PidCache and partition cleanup both rely on receiving termination messages from the remote node since they watch remote actors. However, it doesn't necessarily happen because nodes can be taken down ungracefully and in such a case the termination message never arrives, leaving zombie actors present in the partition (also zombie) and the pid cache, forever.

**Comment by rogeralsing** [comment]
Worth noting is that this is a two step processes.

One node can be the algorithmic owner if an ID, while the actor still resides on another node.
This is to limit downtime when nodes go down, we don't want to move actors that are alive, we just move the location information for them.

**Comment by rogeralsing** [comment]
>stop the transferred actor instead of unwatching it

This would be the wrong approach given the above information.
The node unwatches because it is no longer the owner of the lookup. the actor still belongs to whatever node it was on.

**Comment by rogeralsing** [comment]
The process is described on page 37 and forward in this presentation:
https://github.com/rogeralsing/Presentations/blob/master/FailingGracefully/QConSP/failingactorsflat.pptx

Lets hook up for a chat whenever possible and Ill run through it in detail

**Comment by alexeyzimarev** [comment]
I guess I got confused by three methods called `TransferOwnership`.

Still, when we call `RemoveAddressFromPartition`, all the pids remain in the PidCache. I guess that was the main issue.

Also, when the ownership transfer was there, I saw actors from the dead node still present in the partition. I have an API that calls the partition actor and gets the partition, groups by the address and I clearly saw actors from the dead node still there. Those actors were also in the PidCache and the app was endlessly trying to send to those dead actors.

I agree we need to review it. I made a few more changes, It might all be caused by fast node termination, but I believe that's not an edge case, more like an expected scenario.

**Comment by alexeyzimarev** [comment]
Basically, I changed both `TransferOwnership` and `RemoveAddressFromPartition` by this:

```csharp
        private void EnsureNewAddresses(IContext context)
        {
            foreach (var (actorId, pid) in _partition.ToArray())
            {
                var address = Cluster.MemberList.GetPartition(actorId, _kind);
 
                if (address != pid.Address)
                {
                    RemoveActorsWithChangedAddress(actorId, pid, context);
                }
            }
        }

        private void RemoveActorsWithChangedAddress(string actorId, PID pid, IStopperContext context)
        {
            Logger.LogDebug("[Partition] Removing partition ownership for {Actor}", actorId);

            _partition.Remove(actorId);
            _reversePartition.Remove(pid);
            
            Cluster.PidCache.RemoveCacheByPid(pid);
            
            if (pid.Address == Cluster.System.ProcessRegistry.Address)
                context.Stop(pid);
        }
```

and it seems to work

**Comment by alexeyzimarev** [comment]
No, it doesn't work. The previous version worked better...

**Comment by rogeralsing** [comment]
Also, we should consider this approach:
![image](https://user-images.githubusercontent.com/647031/86837128-c0c93900-c09e-11ea-9228-0a779d870a78.png)

Having this complex cluster code in .net and Go at the same time is problematic.
We already have a Go imp for cluster that use a built in cluster membership protocol.
Allowing C# nodes to piggyback onto this might be a better approach.

**Comment by alexeyzimarev** [comment]
Yeah, I remember you posted it. I just think that my k8s cluster provider works just fine, but the clustering itself is killing me :)

**Comment by rogeralsing** [comment]
the TLDR; is that we have a distributed hashtable. 
this distributed hashtable is responsible for the lookup of where actors are.
there is no affinity between the location of an ID (key into this distributed hash) and the placement of the actors themselves.
This is to prevent moving actors when a new node joins or leaves.
This is also how Orleans do/did it.

A simpler approach could be to do what Akka does, with a cluster singleton that acts as the central repository for ID to placement lookups.

This way, there is always a single source of truth of where an ID lives.
But, the downside is, if this node goes down, everything is dead.

Another thing we can look into now, as the .NET version now have the concept of ActorSystems.
We can easily boot multiple instances from the same .NET process, we could also have some form of in-memory cluster provider.
This would allow us to write proper tests for the cluster w/o spinning up new processes on the test machine.

Even if that doesn't quite reflect a real environment, at least it would allow us to reason and model behaviour better for the cluster concepts, e.g. transfer ownership.

**Comment by rogeralsing** [comment]
Pseudo code:
```csharp
var systems = ClusterTest.StartCluster(nodes: 5);
var sys1 = systems[0];
...
setup stuff
ClusterTest.KillNode(2); //kill node2
...
assert stuff
```

I think this would be pretty neat

**Comment by alexeyzimarev** [comment]
Testing should be quite easy with Docker.

Again, the current issue is that when the node goes down and termination messages to not reach watchers on other nodes, cached addresses stay in the partition cache and reverse cache, as well as the PID cache of the cluster. Some of it was addressed by cleaning up the partition cache from all the actors that belong to the dead node and transferring ownership to the correct partition. But hasn't affected the PID cache and transferred actors also weren't stopped.

I think I solved it in the last iteration. It's a bit dirty since the partition actor has to call the PidCache removal directly, but it works. I can show the code soon.

**Comment by alexeyzimarev** [comment]
I am also thinking of how can we open doors for building something like the Orleans Dashboard. For my tests, I had to make a few internals and privates public so I can watch what is going on via the API. In one case (partition actor) I added a request/response possibility to ask for the partition internals, that's probably a better way of doing this.

Also, bringing it events based on System.Diagnostics could be a good idea. Alternatively, we can publish more events to the event stream and let something that uses System.Diagnostics to subscribe to those events.

**Comment by alexeyzimarev** [comment]
Concerning proxying C# actors via Go nodes, the drawback there is that it will introduce two more network calls for each remote message, which is a bummer.

**Comment by zgramana** [comment]
Definitely not a fan of the Go sidecar approach. This results in greater complexity all around: design, implementation, monitoring, testing, security, performance, etc. It would be a tough sell for me, effectively leaving me, and others I suspect, to roll our own replacements.

I would rather do the work of porting the membership protocol from protoactor-go. If the concern is to avoid hand-porting, there are other approaches that can also extend to support other languages with an incremental effort. For example, I would take prefer generating most, if not all, of the implementation from a model (Drakon, gsl, custom tool, etc) derived from the Go implementation over the sidecar approach.

**Comment by rogeralsing** [comment]
Done in #532
